#Doxygen

##Commenti con marcatore

Se nei commenti aggiungo il marcatore "!", i commenti verranno letti da doxygen e mi permetteranno di comporre la documentazione del mio programma.

* //!
* /*!  */
* \relates "nome classe"
* \pre
* \p

Mi servono anche a formare precondizioni per l'utilizzo dei metodi.

Il marcatore **\relates "nomeclasse"** nei commenti dei metodi all'infuori della classe mi indica che la documentazione di quel metodo sarà correlata a quella della classe.

Il marcatore **\pre** mi permette di definire delle precondizioni che verrano descritte nella documentazione.

Il marcatore **\p** permette di indicare la prossima parola come un parametro che verrà formattata diversamente nella documentazione.

Doxygen funziona come un compilatore, permette di creare documentazione C++ interpretando il codice e tutti i commenti che hanno i vari marcatori.

---

Le assert, oltre che all'inizio e alla fine di un algoritmo, posso inserirle anche durante le varie operazioni. 

Faccio così per migliorare il debugging e per capire facilmente dove sono eventuali errori (sopratutto se un algoritmo è diviso in più fasi o è complicato).

---

Per migliorare le prestazioni, posso creare un altro header file che conterrà tutte le definizioni delle funzioni inline.

Se definisco la funzione all'interno della classe, è implicitamente definita come inline.

---

#Contratto Wide

Per segnalare errori nei costruttori utilizzo le eccezioni

    Razionale::Razionale(Int num, Int den)
    {
      if(den==0)
        throw Razionale_Invalido();
      ...
      ...
    }

Lanciando l'eccezione nel costruttore mi permette di uscire da esso senza creare l'oggetto, quindi non verrà allocata memoria e non verrà chiamato il distruttore.

##Esempio:

    void job()
    {
      Res* r1 = acquisisci_risorsa("res1");
      Res* r2 = acquisisci_risorsa("res2");
      do_task(r1,r2);
      rilascia_risorsa(r2);
      Res* r3 = acquisisci_risorsa("res3");
      do_task(r1,r3);
      rilascia_risorsa(r3);
      rilascia_risorsa(r1);
    }

Se utilizzo una risorsa e non mi serve più devo sempre rilasciarla per evitare **resource leak**.

Il flusso della funzione non è unico in quanto ogni volta che richiamo una funzione essa può generare eccezioni che causano il salto delle istruzioni fino alla fine della funzione. 

Il flusso quindi può avere numerose biforcazioni in base al numero di chiamate di funzioni all'interno della funzione.

* Considerando numerosi flussi di esecuzione, sono sicuro di non avere resource leak?

Se sono completamente sicuro che tutto va a buon fine non ho resource leak.

Contrariamente se qualche funzione mi lancia un'eccezione non sono così sicuro di non avere resource leak (a meno che l'eccezione non la lanciano le funzioni che acquisiscono le risorse, in quel caso le risorse non vengono neanche acquisite).

In poche parole il codice è *safe* se suppongo l'esecuzione senza problemi, è *unsafe* se prevedo delle eccezioni nell'esecuzione. Dico che è *exception safe* se il codice rilascia tutte le risorse anche se vengono sollevate delle eccezioni.

* Come faccio ad avere un codice *exception safe*?

Utilizzo dei try per ogni risorsa che viene acquisita, metto dei catch(...) che catturano qualsiasi errore e nel catch rilascio la risorsa rilanciando l'eccezione

    void job()
        {
          Res* r1 = acquisisci_risorsa("res1");
		  try {  //Protettore di r1
            
			Res* r2 = acquisisci_risorsa("res2");
			try {   //Protettore di r2
    	      do_task(r1,r2);
    	      rilascia_risorsa(r2);
      	    }
			catch (...) {
			  rilascia_risorsa(r2);
			  throw;
			} 
			rilascia_risorsa(r1);
            rilascia_risorsa(r2);
            Res* r3 = acquisisci_risorsa("res3");
			try {   //Protettore di r3
              do_task(r1,r3);
              rilascia_risorsa(r3);
              rilascia_risorsa(r1);
			}
			catch (...) {
			  rilascia_risorsa(r3);
			  throw;
		    }
		  }
		  catch (...) {
		    rilascia_risorsa(r1);
			throw;
        }
		
		
* Throw senza parametri può essere lanciata solo da dentro un'eccezione e mi rilancia l'eccezione catturata, utilizza la neutralità rispetto alle eccezioni. È una caratteristica fondamentale per avere codice exception safe
* Metodo troppo verboso, ripetitivo e può generare numerosi errori. Viene infatti usato dai barbari dei programmatori java che però invece del catch utilizzano il finally che rilancia automaticamente l'eccezione senza riscrivere throw.
* Finally non è presente in C++ perchè c'è un'alternativa migliore che RAII - RRID :D

#RAII - RRID
###Resource Acquisition Is Initialization
###Resource Release Is Destruction

    class RAII_RRID_Res {
	public:
	  RAII_RRID_Res(const char* name) {
	    r = acquisisci_risorsa(name);
	  }
	  ~RAII_RRID_Res(){
	    rilascia_risorsa(r);
	  }
	  
	  //Costruttori di copia e spostamento
	  RAII_RRID_Res(const RAII_RRID_Res&) = delete;
	  RAII_RRID_Res& operator=(const RAII_RRID_Res&) = delete;
	  //Facendo così disabilito questi due costruttori
	  
	private:
	  Res* r;
	};  
	
	void job()
    {
	  RAII_RRID_Res r1("res1");
	  {
	    RAII_RRID_Res r2("res2");
        do_task(r1,r2);
		//Creo uno scope così elimino res2 quando mi pare ammé
	  }
	  RAII_RRID_Res r3("res3");
      do_task(r1,r3);
	}

* Tolgo le funzioni "rilascia_risorsa()" perchè, essendo le risorse oggetti della classe istanziati, quando finisce lo scope, il distruttore della classe viene chiamato e distrugge le risorse implicitamente.
* Il codice si rimpiccilisce, è più leggibile e non si ci può dimenticare di deallocare le risorse.

Idioma RAII = costrutto linguistico che si applica solo in un particolare linguaggio e che non si può utilizzare in altri in quanto la traduzione non è possibile (come in java perchè non ha i costruttori).

La classe di prima ha un problema, non ho modo di utilizzare le risorse perchè non ho un puntatore alla classe. Quindi:

* Uso un getter
* Faccio conversione implicita: ``operator Res*() const {return r};``

Mancano anche le definizioni dei costruttori di copia e di spostamento che possono accedere ai membri anche se sono stati distrutti andando in *undefined behavoir*. Quindi se definisco un costruttore, se voglio fare le cose per bene, devo dichiarare anche gli altri costruttori (modifico il codice sopra mica lo riscrivo ;) )


